timely_beliefs/sensors/func_store/knowledge_horizons.py

```diff
@@ -31,6 +34,39 @@ def at_date(
     return event_start - knowledge_time.astimezone(event_start.tzinfo)
 
 
+def x_years_ago_at_date(
+    event_start: datetime | pd.DatetimeIndex,
+    day: int,
+    month: int,
+    x: int = 1,
+    get_bounds: bool = False,
+    z: str | None = None,
+) -> timedelta | pd.TimedeltaIndex | tuple[timedelta, timedelta]:
+    """Compute the sensor's knowledge horizon to represent the event could be known since some date, `x` years ago.
+
+    For example, it can be used for a tax rate that changes annually and with a known publication date.
+
+    :param event_start:     Start of the event, used as an anchor for determining the knowledge horizon.
+    :param day:             Reference day of the month of the annual date to compare against.
+    :param month:           The month of the annual date to compare against.
+    :param x:               The number of years to shift the reference date to.
+    :param z:           Timezone string.
+    :param get_bounds:      If True, this function returns bounds on the possible return value.
+                            These bounds are normally useful for creating more efficient database queries when filtering by belief time.
+    """
+    if get_bounds:
+        return timedelta.min, timedelta.days(366 * x)
```

Comment:
I believe the bounds should be `timedelta(days=(x-1)*366+1), timedelta(days=(x+1)*366)`, assuming x>1!

Explanation for x=1:

The smallest possible value happens with an event at midnight January 1st and a knowledge time at midnight 31 December the preceding year. -> 1 day

The largest value happens with an event just before midnight December 31st and a knowledge time at midnight January 1st the preceding year. -> nearly 2 years

Reply:
NB: I reckon things like tax returns concerning previous years may lead to sensors with negative x, so that still needs some thinking (or otherwise explicitly raising a `NotImplementedError` for x<=0).

Reply:
Thanks for the explanation.
I've added both the bounds and the value check.